### ISSUE TITLE
Remove `g++` dependency to find out the system headers

### ISSUE THREAD

**Issue by JBakamovic** [JBakamovic/cxxd#2]
Solution:
1. (**Better**) Find out which toolchain is being used for the project given the configuration file we are using.
2. Depending on the platform we're running on use appropriate toolchain. But then we have to find out which ones are existing on the system and we can run into another bunch of problems ...

**Comment by MaskRay** [comment]
It is available as `clang::HeaderSearchOptions::UserEntries` https://github.com/llvm-mirror/clang/tree/master/include/clang/Lex/HeaderSearchOptions.h#L102

But it isn't exposed as a libclang API.

`g++/clang++ -v -E -x c++` is the best for a libclang based tool.

**Comment by JBakamovic** [comment]
> g++/clang++ -v -E -x c++ is the best for a libclang based tool.

Yes, that's exactly how it is implemented [right now](https://github.com/JBakamovic/cxxd/blob/master/parser/compiler_args.py#L142-L150) but it would be nice to have this information exposed by `libclang` API as you mentioned. It probably shouldn't be too hard to write a patch for that? I may check it out if I find some spare time ... I'm currently occupied with finishing the code-completion feature.

Although, having a dependency on a C/C++ compiler for C/C++ language server is not to be unexpected, right :smiley:

**Comment by MaskRay** [comment]
It would be hard. As you may probably end up replicating lots of C++ types in libclang counterparts. I am unclear if this kind of change will be accepted by upstream clang.

Another thing that may be out-of-scope of this function is when cross compilation is involved. clangDriver also has logic to detect various system search paths when it knows it is used as cross compiler, e.g. `clang --target=ppc64le` `aarch-linux-gnu-clang`

**Comment by JBakamovic** [comment]
> It would be hard. As you may probably end up replicating lots of C++ types in libclang counterparts. I am unclear if this kind of change will be accepted by upstream clang.

I am not sure I see why that would be the case. For example, `CompilerInstance` header is [already used by the `libclang` implementation](https://github.com/llvm-mirror/clang/blob/master/tools/libclang/CIndex.cpp#L33) which is AFAICT exactly what we would need to implement fetch-system-headers functionality: there is [`getHeaderSearchOpts()`](https://github.com/llvm-mirror/clang/blob/master/include/clang/Frontend/CompilerInstance.h#L299-L301) in `CompilerInstance` class.

> Another thing that may be out-of-scope of this function is when cross compilation is involved. clangDriver also has logic to detect various system search paths when it knows it is used as cross compiler, e.g. clang --target=ppc64le aarch-linux-gnu-clang

Cross-compilations are usually handled by setting up proper `sysroot` and seems like this is also [configurable via `Sysroot` field from `HeaderSearchOptions`](https://github.com/llvm-mirror/clang/blob/master/include/clang/Lex/HeaderSearchOptions.h#L206). As far as I am concerned I believe this is already handled by internal clang driver implementation.
